storage/store.go

```diff
@@ -1198,97 +1198,100 @@ func (s *Store) Descriptor() (*proto.StoreDescriptor, error) {
+
+	if !argsHeader.Timestamp.Equal(proto.ZeroTimestamp) {
 		if s.Clock().MaxOffset() > 0 {
 			// Once a command is submitted to raft, all replicas' logical
 			// clocks will be ratcheted forward to match. If the command
 			// appears to come from a node with a bad clock, reject it now
 			// before we reach that point.
-			offset := time.Duration(header.Timestamp.WallTime - s.Clock().PhysicalNow())
+			offset := time.Duration(argsHeader.Timestamp.WallTime - s.Clock().PhysicalNow())
 			if offset > s.Clock().MaxOffset() {
-				err := util.Errorf("Rejecting command with timestamp in the future: %d (%s ahead)",
-					header.Timestamp.WallTime, offset)
-				reply.Header().SetGoError(err)
-				return err
+				err = util.Errorf("Rejecting command with timestamp in the future: %d (%s ahead)",
+					argsHeader.Timestamp.WallTime, offset)
+				return
 			}
 		}
 		// Update our clock with the incoming request timestamp. This
 		// advances the local node's clock to a high water mark from
 		// amongst all nodes with which it has interacted.
-		s.ctx.Clock.Update(header.Timestamp)
+		s.ctx.Clock.Update(argsHeader.Timestamp)
 	}
 
+	ctx = s.Context(ctx)
+
 	// Backoff and retry loop for handling errors.
 	retryOpts := *s.ctx.RangeRetryOptions
 	retryOpts.Tag = fmt.Sprintf("store: %s", args.Method())
-	err := retry.WithBackoff(retryOpts, func() (retry.Status, error) {
+	err = retry.WithBackoff(retryOpts, func() (retry.Status, error) {
 		// Add the command to the range for execution; exit retry loop on success.
 		reply.Reset()
 
 		// Get range and add command to the range for execution.
-		rng, err := s.GetRange(header.RaftID)
+		rng, err := s.GetRange(argsHeader.RaftID)
 		if err != nil {
-			reply.Header().SetGoError(err)
-			return retry.Break, err
-		}
-
-		if err = rng.AddCmd(ctx, proto.Call{Args: args, Reply: reply}, true); err == nil {
 			return retry.Break, err
 		}
 
-		// Maybe resolve a potential write intent error. We do this here
-		// because this is the code path with the requesting client
-		// waiting. We don't want every replica to attempt to resolve the
-		// intent independently, so we can't do it in Range.executeCmd.
-		if wiErr, ok := err.(*proto.WriteIntentError); ok {
-			pushType := proto.PUSH_TIMESTAMP
-			if proto.IsWrite(args) {
-				pushType = proto.ABORT_TXN
+		if err = rng.AddCmd(ctx, proto.Call{Args: args, Reply: reply}, true); err != nil {
+			// Maybe resolve a potential write intent error. We do this here
+			// because this is the code path with the requesting client
+			// waiting. We don't want every replica to attempt to resolve the
+			// intent independently, so we can't do it in Range.executeCmd.
+			if wiErr, ok := err.(*proto.WriteIntentError); ok {
+				var pushType proto.PushTxnType
+				if proto.IsWrite(args) {
+					pushType = proto.ABORT_TXN
+				} else {
+					pushType = proto.PUSH_TIMESTAMP
+				}
+				err = s.resolveWriteIntentError(ctx, wiErr, rng, args, pushType, false /* wait */)
```

Comment:
The old code would clear reply.Header().GoError if resolveWriteIntentError returns nil. I don't think resolveWriteIntentError *can* return nil, though.